### ISSUE TITLE
Q: How does CVE to package matching work?

### ISSUE THREAD

**Issue by terotil** [repology/repology-webapp#230]
Tried to figure out how do CVEs (CPEs) map to packages. Table `manual_cpes` seems to be a mapping from CPE to package, so does `project_cpe`. Also `packages` has full CPE fieldset, i.e. provides the mapping by itself. Which one is the authoritative and what's the role of others?

I'm not working on anything in repology. Just curious about how package identifiers work in the wild.

**Comment by AMDmi3** [comment]
> Tried to figure out how do CVEs (CPEs) map to packages. Table `manual_cpes` seems to be a mapping from CPE to package, so does `project_cpe`.

Only `manual_cpes` is used, `projects_cpe` is currently not used for anything. Before, these both were used through `all_cpes` view which unions them:

https://github.com/repology/repology-updater/commit/3889d432de082f6c10c6e4dec7f1e48e3cd5ae54#diff-4e29f1e9506a3b50edfc7c875b11516b44e7b95aac2d232b977424ddada05a68L850

but cpe data from packages was deemed inaccurate. For the record, I haven't updated manual cpes for quite some time, so it's not accurate (or rather, complete) either.

> Also `packages` has full CPE fieldset, i.e. provides the mapping by itself. Which one is the authoritative and what's the role of others?

`packages` is raw data parsed from repositories, it's huge and is not directly usable for most purposes, but you may call it ground truth. Most other tables including `project_cpe` contains data slices extracted from it and kept in sync through updates.

**Comment by terotil** [comment]
Thanks for quick response and detailed insight!

So all the data in `manual_cpes` is literally you typing manually new rows in the table as needed? :-O

**Comment by AMDmi3** [comment]
> So all the data in `manual_cpes` is literally you typing manually new rows in the table as needed? :-O

Not so hardcore. There's admin web UI for this task in repology-webapp - it lists latest unmatched CVEs and suggests projects they could be linked to, so in many cases it's just pressing a bunch of buttons. Still it takes time to verify all these suggestions.

**Comment by terotil** [comment]
> admin web UI [...] lists latest unmatched CVEs and suggests projects

Found my way to https://github.com/repology/repology-webapp/blob/c06560a49b896805a8ed60cbbfb5bd934ebef026/sql.d/cpes/get_recent_cve_misses.sql#L80 and saw suggestions having `effname` matching `cpe_product` with or without `cpe_target_sw`.

I'm probably assuming correctly that there's no package to CPE mapping for packages, which never had any CVEs issued on them.

**Comment by AMDmi3** [comment]
> I'm probably assuming correctly that there's no package to CPE mapping for packages, which never had any CVEs issued on them.

Yes. Well, technically nothing prevents that, there's also webui to add CPEs completely manually, but I don't see how that can be used without CVEs, as NVD CPEs are practically unpredictable. If there's a way to discover CPE vendor/project for a project without any CVEs published, and we can assume that NVD will use the same ids, there's no problem in adding these to repology.

**Comment by fthdrmzzz** [comment]
Hello,

I am currently working on a security tool and I have came across this project (which is great).

For the security tool I am working on, I need to find matches between CPEs and the packages. I have downloaded repology bulk data to see if I can make use of it but I couldn't understand from which table I can get the matchings from. How can I retrieve that data?

when I checked `packages` I saw that columns containing CPE data is empty. Is there a repology app that I need to run to make matches?

**Comment by AMDmi3** [comment]
See https://github.com/repology/repology-webapp/blob/master/sql.d/get_project_cves.sql

> when I checked `packages` I saw that columns containing CPE data is empty

Thats CPE (usually vendor:product:* masks) which some repositories provide. There are not many repositories which do and these are not reliable anyway.

**Comment by fthdrmzzz** [comment]
Okay, so you filter the CVEs using vendor-product pairs on `manual_cpes` table. I understand that repology doesn't claim to be a security product. But I want to ask just to understand. 

- Application type CPEs, (i.e `cpe_dictionary` table in repology) contains around 40k records. and around 36k are unique by product name.
- `manual_cpes` contains around ~3.8k records. 

Assuming `manual_cpes` bindings are valid, we are missing around 32k cpe bindings for packages right?

**Comment by AMDmi3** [comment]
> Assuming `manual_cpes` bindings are valid, we are missing around 32k cpe bindings for packages right?

Not on that scale. Most missing CPE bindings are for proprietary products which are unlikely to be packaged:

```
repology=> select cpe_vendor, count(*) from (select distinct cpe_vendor, cpe_product from cpe_dictionary) as _ group by cpe_vendor order by count desc;
     cpe_vendor     | count 
--------------------+-------
 ibm                |   837
 oracle             |   829
 hp                 |   701
 cisco              |   634
 jenkins            |   629  <- mostly plugins, not packaged
 intel              |   569
 microsoft          |   568
 redhat             |   392
 sap                |   354
 ricoh              |   311
 adobe              |   304
 apache             |   301  <- some real projects here
 siemens            |   287
 broadcom           |   266
 dell               |   240
 altiris            |   211
 symantec           |   207
 vmware             |   186
 citrix             |   147
```

There are other large categories not relevant for Repology for the same reason, such as plugins or node/javascript modules. We also probably don't need historic entries beyond repository long term support periods.

But in general you are right - CVE coverage depends on manual maintenance of CPE bindings and I'm not very responsive about that. Updates are visible [here](https://repology.org/security/recent-cpes) and you can see that I've missed about half a year worth of updates after Sep 2023. It's still possible to catch up though.

But despite that, I would estimate Repology's CPE coverage as "fine", judging by the rate of new unmatched CVEs which is quite low, as all the most CVE-active projects are already CPE-covered and thus new CVEs for these are automatically matched. I think the database contains enough data to measure this more precisely.

Note also that NVD is not very active lately, citing [announcement](https://nvd.nist.gov/general/news/nvd-program-transition-announcement):
> There is a growing backlog of vulnerabilities submitted to the NVD and requiring analysis. This is based on a variety of factors, including an increase in software and, therefore, vulnerabilities, as well as a change in interagency support.
>
> Currently, we are prioritizing analysis of the most significant vulnerabilities

so even if Repology's coverage of NVD is fine, NVD's coverage of vulnerabilities in the wild may not be, so if you're interested in using Repology for security purposes, it may not be reliable. If there are other CVE sources (I'm not aware but I haven't done any research), we may investigate possibility of adding them.

**Comment by fthdrmzzz** [comment]
Hello! Thank you for your answers so far :)

I have another question. Is there a reason for not using OVAL that is maintained by distributions to do package CVE matches? Or is it being used already?

**Comment by AMDmi3** [comment]
I have no idea that that is and I don't think I have encountered this abbreviation in any repository data.

**Comment by fthdrmzzz** [comment]
Okay, It stands for (O)pen Source (V)ulnerabilitiy (A)ssessment (L)anguage. It is a standart developed by MITRE and adopted by communities. What I understood is, packages with their versions and corresponding CVEs are documented after the responsible people assessing it. 

Here is an example record I got from Ubuntu Bionic Oval document:
```xml
    <definition class="vulnerability" id="oval:com.ubuntu.bionic:def:18040156300000" version="1">
      <metadata>
        <title>openconnect</title>
        <reference source="Package" ref_id="openconnect" ref_url="https://launchpad.net/ubuntu/+source/openconnect"/>
        <description>An SSL VPN client</description>
        <affected family="unix">
          <platform>Ubuntu 18.04 LTS</platform>
        </affected>
        <advisory>
          <rights>Copyright (C) 2024 Canonical Ltd.</rights>
          <component>universe</component>
          <current_version>7.08-3ubuntu0.18.04.2</current_version>
          <cve href="https://ubuntu.com/security/CVE-2013-7098" priority="medium" public="20200213" cvss_score="9.8" cvss_vector="CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H" cvss_severity="critical" test_ref="oval:com.ubuntu.bionic:tst:18040156300000">CVE-2013-7098</cve>
          <cve href="https://ubuntu.com/security/CVE-2019-16239" priority="medium" public="20190917" cvss_score="9.8" cvss_vector="CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H" cvss_severity="critical" usns="4565-1,4565-1" test_ref="oval:com.ubuntu.bionic:tst:18040156300010">CVE-2019-16239</cve>
        </advisory>
      </metadata>
      <criteria>
        <extend_definition definition_ref="oval:com.ubuntu.bionic:def:100" comment="Ubuntu 18.04 LTS is installed." applicability_check="true"/>
        <criteria operator="OR">
          <criterion test_ref="oval:com.ubuntu.bionic:tst:18040156300000" comment="(CVE-2013-7098) openconnect package in bionic, is related to the CVE in some way and has been fixed (note: '5.02-1')."/>
          <criterion test_ref="oval:com.ubuntu.bionic:tst:18040156300010" comment="(CVE-2019-16239) openconnect package in bionic, is related to the CVE in some way and has been fixed (note: '7.08-3ubuntu0.18.04.2')."/>
        </criteria>
      </criteria>
    </definition>
```

One possible disadvantage could be, systems-distros that doesn't have strong communities might not have an OVAL document. But other than that, I think this can be very useful.  (I am sad because I discovered this at the end of my thesis duration :) but thats ok)

**Comment by AMDmi3** [comment]
Oh, that. I believe there are multiple formats, for instance FreeBSD uses vuxml https://github.com/freebsd/freebsd-ports/blob/main/security/vuxml/vuln/2024.xml for seemingly similar stuff. These could be used to clear vulnerable status for patched packages and probably for a small fraction of vulnerabilities not otherwise known, but these seem hard to handle (for instance, these use distro/os specific version comparison algorithms).
